src/group_theory/order_of_element.lean

```diff
@@ -103,40 +103,203 @@ lemma card_quotient_dvd_card (s : subgroup α) [decidable_pred (λ a, a ∈ s)]
   fintype.card (quotient s) ∣ fintype.card α :=
 by simp [card_eq_card_quotient_mul_card_subgroup s]
 
-lemma exists_gpow_eq_one (a : α) : ∃i≠0, a ^ (i:ℤ) = 1 :=
-have ¬ injective (λi:ℤ, a ^ i),
-  from not_injective_infinite_fintype _,
-let ⟨i, j, a_eq, ne⟩ := show ∃(i j : ℤ), a ^ i = a ^ j ∧ i ≠ j,
-  by rw [injective] at this; simpa [not_forall] in
-have a ^ (i - j) = 1,
-  by simp [sub_eq_add_neg, gpow_add, gpow_neg, a_eq],
-⟨i - j, sub_ne_zero.mpr ne, this⟩
-
-lemma exists_pow_eq_one (a : α) : ∃i > 0, a ^ i = 1 :=
-let ⟨i, hi, eq⟩ := exists_gpow_eq_one a in
+end order_of
+
+section order_of
+
+section monoid
+variables {α} [monoid α]
+variables (a) [decidable_pred (λ n, 0 < n ∧ a ^ n = 1)] [decidable (∃ n, 0 < n ∧ a ^ n = 1)]
+
+/-- `order_of a` is the order of the element `a`, i.e. the `n ≥ 1`, s.t. `a ^ n = 1` if it exists.
+Otherwise, i.e. if `a` is of infinite order, then `order_of a` is `0` by convention.-/
+def order_of  : ℕ :=
+if h : ∃ n, 0 < n ∧ a ^ n = 1 then nat.find h else 0
+
+variable {a}
+
+lemma order_of_of_finite_order (h : ∃ n, 0 < n ∧ a ^ n = 1) : order_of a = nat.find h :=
+begin
+  rw order_of,
+  split_ifs with hx,
+  refl
+end
+
+variable (a)
+
+lemma pow_order_of_eq_one : a ^ order_of a = 1 :=
+begin
+  rw order_of,
+  split_ifs with hx,
+  { exact (nat.find_spec hx).2 },
+  { exact pow_zero a }
+end
+
+variable {a}
+
+lemma order_of_pos (h : ∃ n, 0 < n ∧ a ^ n = 1) : 0 < order_of a :=
+begin
+  rw order_of_of_finite_order h,
+  exact (nat.find_spec h).1
+end
+
+lemma order_of_le_of_pow_eq_one {n : ℕ} (hn : 0 < n) (h : a ^ n = 1) : order_of a ≤ n :=
+begin
+rw order_of,
+split_ifs with ha,
+  { exact nat.find_min' ha ⟨hn, h⟩ },
+  { exact le_of_lt hn }
+end
+
+instance decidable_order_of_one : decidable (∃ n, 0 < n ∧ (1 : α) ^ n = 1) :=
+decidable.is_true ⟨1, ⟨nat.one_pos, one_pow 1⟩⟩
+
+instance decidable_pred_order_of_one : decidable_pred (λ n, 0 < n ∧ (1 : α)  ^ n = 1) :=
+begin
+  simp only [one_pow, and_true, eq_self_iff_true],
+  exact set.decidable_set_of (λ (a : ℕ), 1 ≤ a)
+end
```

Comment:
I'm pretty sure that your `variables` statement is leaking into these definitions; `#lint` should help find out.

Reply:
`#lint` doesn't report any problems.